rust/src/pet/message/sum.rs

```diff
@@ -0,0 +1,147 @@
+#![allow(dead_code)] // temporary
+
+use std::ops::Range;
+
+use sodiumoxide::crypto::{box_, sign};
+
+use super::{MsgBoxBufMut, MsgBoxBufRef, MsgBoxDecr, MsgBoxEncr, SUM_TAG};
+use crate::pet::PetError;
+
+// sum box field ranges
+const EPHM_PK_RANGE: Range<usize> = 65..97; // 32 bytes
+
+/// Mutable and immutable buffer access to sum box fields.
+struct SumBoxBuffer<B> {
+    bytes: B,
+}
+
+impl SumBoxBuffer<Vec<u8>> {
```

Comment:
nit: I don't know if it's really necessary to have an impl for `Vec<u8>` specifically. Maybe:

```rust
impl<B> for SumBoxBuffer<B> {
    fn new(bytes: B) -> Self {
        Self { bytes }
    }
}
```

would be enough? That's personal taste though, so feel free to ignore.

Reply:
with a generic buffer of type `B` there is no guarantee that any of the accessors can actually be used. what i would like to express is that `B` should be trait-bounded by at least one of the bounds `AsRef<[u8]> + ?Sized` or `AsMut<[u8]>`, but i'm not sure how to do this without having separate methods `new` and `new_mut` which i would like to avoid.

therefore i went with the type specific `Vec<u8>` impl. not sure what's the best approach here.

Reply:
No need for any trait bound. If someone makes a buffer out of something that does not impl `AsRef<[u8]>` or `AsMut<[u8]>`, and then tries to use a getter or setter, the compiler is going to yell at them, that the method is not implemented for that type.

Reply:
i just noticed that we are mixing up two different impl here. the `from` method is already generic over `B`:
```
impl<B: AsRef<[u8]>> SumBoxBuffer<B> {
    /// Create a sum box buffer from `bytes`. Fails if the `bytes` don't conform to the expected
    /// sum box length `exp_len`.
    fn from(bytes: B, exp_len: usize) -> Result<Self, PetError> {
        (bytes.as_ref().len() == exp_len)
            .then_some(Self { bytes })
            .ok_or(PetError::InvalidMessage)
    }
}
```
whereas the `new` method cannot be generic over `B` since it has no input (aside from the length parameter):
```
impl SumBoxBuffer<Vec<u8>> {
    /// Create an empty sum box buffer of size `len`.
    fn new(len: usize) -> Self {
        Self {
            bytes: vec![0_u8; len],
        }
    }
}
```